File: Server/src/dummyClient/serverTime.go
package main

import (
	"fmt"
	"time"
)

type ServerTime struct {
	serverTime int64
	diffTime   int64
	rttTime    float64
}

var alpha float64 = 0.5

func (st *ServerTime) OnRecvSyncTime(serverTime int64) {
	st.serverTime = serverTime
	st.diffTime = time.Now().UnixNano() - serverTime
}

func (st *ServerTime) OnRecvPong(clientTime int64, serverTime int64) {
	currentTime := time.Now().UnixNano()
	st.serverTime = serverTime
	st.diffTime = currentTime - serverTime

	fmt.Println("OnRecvPong currentTime:", currentTime, " clientTime:", clientTime)

	if st.rttTime == 0 {
		st.rttTime = float64(currentTime-clientTime) / float64(time.Millisecond)
	} else {
		st.rttTime = alpha*st.rttTime + (1-alpha)*(float64(currentTime-clientTime)/float64(time.Millisecond))
	}
}

func (st *ServerTime) GetServerTime(clientTime int64) int64 {
	return clientTime - st.diffTime + int64(st.rttTime/2*float64(time.Millisecond))
}

func (st *ServerTime) GetClientTime(serverTime int64) int64 {
	return serverTime + st.diffTime - int64(st.rttTime/2*float64(time.Millisecond))
}

File: UnityProject/SHU/Assets/Scripts/SHU/Network/ServerTime.cs
using System;
using UnityEngine;

namespace SHU {
  namespace Network {
    public class ServerTime : MonoBehaviour {
      long serverTime;
      long diffTime;
      float rttTime;
      float alpha = 0.5f;
      const long MilliToNano = 1000000;
      private System.Object thisLock = new System.Object();

      public void OnRecvTimePacket(PacketData data) {
        lock (thisLock) {
          if (data.packetID == PacketID.SyncTime) {
            serverTime = (long)(BitConverter.ToUInt64 (data.data, 0));
            diffTime = GetNowUnixTime () - serverTime;
          } else if (data.packetID == PacketID.Pong) {
            long clientTime = (long)(BitConverter.ToUInt64 (data.data, 0));
            serverTime = (long)(BitConverter.ToUInt64 (data.data, 8));

            long currentTime = GetNowUnixTime ();
            diffTime = currentTime - serverTime;

            if (rttTime == 0) {
              rttTime = (float)((currentTime - clientTime)/MilliToNano);
            } else {
              rttTime = alpha * rttTime + (1 - alpha) * (float)((currentTime - clientTime)/MilliToNano);
            }

            Debug.Log (string.Format ("Ping RTT={0}, Diff={1}", rttTime, diffTime));
          }
        }
      }

      public static long GetNowUnixTime() {
        return (long)(DateTime.UtcNow.Subtract (new DateTime (1970, 1, 1)).TotalMilliseconds * MilliToNano);
      }

      public long GetServerTime(long clientTime) {
        return clientTime - diffTime + (long)(rttTime / 2 * MilliToNano);
      }

      public long GetClientTime(long serverTime) {
        return serverTime + diffTime - (long)(rttTime / 2 * MilliToNano);
      }
    }
  }
}

